In [ ]:
import numpy as np
import pandas as pd

In [ ]:
# path of data

data_path = "/content/smartphones.csv"

df = pd.read_csv(data_path)

df.head()

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,OnePlus 11 5G,"₹54,999",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,Samsung Galaxy A14 5G,"₹16,499",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13
3,Motorola Moto G62 5G,"₹14,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
4,Realme 10 Pro Plus,"₹24,999",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...,Memory Card Not Supported,Android v13


2 types of Data Cleaning ideologies:

1. Noisy Data/ Dirty Data- Useless data, missing values.
2. Messy data - Useful Data but unstructured. --> Structure

### Major types of data cleanliness issues:


1. **Completeness Issues** – Data is missing or incomplete where it is expected to be present.

2. **Validity Issues** – Data doesn't conform to defined formats, types, or rules (e.g., age = -5).

3. **Consistency Issues** – Conflicting data exists across different systems or columns (e.g., birthdate vs. age mismatch).

4. **Accuracy Issues** – Data does not reflect the real-world values correctly (e.g., wrong addresses or names).

5. **Uniqueness Issues** – Duplicate or repeated entries exist where only one should be present.

---


In [ ]:
# shape of the data

df.shape

(1020, 11)

In [ ]:
# check for missing values

df.isna().sum()

,0
model,0
price,0
rating,141
sim,0
processor,0
ram,0
battery,0
display,0
camera,1
card,7


In [ ]:
(
    df
    .isna()
    .any(axis=1)
)

,0
0,False
1,False
2,False
3,False
4,False
...,...
1015,False
1016,False
1017,False
1018,False


In [ ]:
# check for rows having missing values

df.loc[(
    df
    .isna()
    .any(axis=1)
),:]

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
14,Samsung Galaxy S23 Ultra 5G,"₹1,14,990",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 256 GB inbuilt",5000 mAh Battery with 45W Fast Charging,"6.8 inches, 1440 x 3088 px, 120 Hz Display wit...",200 MP Quad Rear & 12 MP Front Camera,Memory Card Not Supported,Android v13
29,OnePlus 11 Pro,"₹69,999",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Android v13,No FM Radio
37,Samsung Galaxy S22 Ultra 5G,"₹91,999",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen1, Octa Core, 3 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 45W Fast Charging,"6.8 inches, 1440 x 3080 px, 120 Hz Display wit...",108 MP Quad Rear & 40 MP Front Camera,Android v12,Bluetooth
49,Samsung Galaxy A74 5G,"₹42,999",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 7 Gen1, Octa Core, 2.36 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP Quad Rear & 32 MP Front Camera,"Memory Card (Hybrid), upto 1 TB",Android v12
69,Oppo Find N Fold,"₹99,990",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","8 GB RAM, 256 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"7.1 inches, 1792 x 1920 px, 120 Hz Display wit...","Foldable Display, Dual Display",64 MP + 10 MP + 8 MP Triple Rear & 32 MP Front...,Android v12
...,...,...,...,...,...,...,...,...,...,...,...
996,Vivo Y55S,"₹13,490",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Snapdragon 425, Quad Core, 1.2 GHz Processor","3 GB RAM, 16 GB inbuilt",2730 mAh Battery,"5.2 inches, 720 x 1280 px Display",13 MP Rear & 5 MP Front Camera,Memory Card Supported,Android v6.0 (Marshmallow)
1000,XTouch F40 Flip,"₹1,999",NaN,Dual Sim,No 3G,No Wifi,"32 MB RAM, 32 MB inbuilt",800 mAh Battery,"1.77 inches, 240 x 320 px Display",Dual Display,1.3 MP Rear Camera
1003,Lava X3,"₹6,999",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio A22, Quad Core, 2 GHz Processor","3 GB RAM, 32 GB inbuilt",4000 mAh Battery,"6.5 inches, 1600 x 720 px Display with Water D...",8 MP Dual Rear & 5 MP Front Camera,"Memory Card (Hybrid), upto 512 GB",Android v12
1012,itel A23s,"₹4,787",NaN,"Dual Sim, 3G, 4G, Wi-Fi","Spreadtrum SC9832E, Quad Core, 1.4 GHz Processor","2 GB RAM, 32 GB inbuilt",3020 mAh Battery,"5 inches, 854 x 480 px Display",2 MP Rear Camera,Android v11,No FM Radio


In [ ]:
# duplicate data

df.duplicated().sum()

np.int64(0)

## Overall Observations

- There are around 1020 rows and 11 columns in the data.
- There are no duplicate rows in the data.
- There are a few missing values in the data in columns:
    1. `rating`
    2. `camera`
    3. `card`
    4. `os`
- Remove feature phones from the data.

In [ ]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

## Column Wise Observations

### Model

- The model column has two pieces of info.
    1. Brand Name
    2. Model Name
- Row 15, 45, 50, 52 We have Ram and storage(in paranthesis) info along with model info - `Validation`

### Price
- Rupee sign - `Validation`
- Commas in values - `Validation`

### Rating
- Missing values in column - 141 - `Completeness`
- Ratings should be on the scale of 10 - `Consistency`

### Sim
- We have to see the number of SIM's supported.
- Bands Supported
- Has Wifi
- Has Extra features - (NFC, IR Blaster)

### Processor
- Processor name
    1. Brand Name
    2. Model Name
- Number of Cores - Str --> Integers
- CPU speed given in GHz

#### Row wise
    - Row 120- tesla phone - Here only number of cores are given - `accuracy`
    - Row 154 - Only Processor name is given - `consistency`
    - Row 203 - Processor name is not given - `consistency`
    - Row 584 - Data validation and inaccurate

### RAM
- The first piece of info is about RAM.(in GB)
- The second is about storage.(in GB)

- Apple phone has only storage info.

### Battery
1. Size of the battery.
2. Fast charging present or not.
3. If FC then wattage.

### Display
1. Display size.
2. Resolution
3. Refresh Rate
4. Display shape

- Apple phones have data shifting issues.

### Camera
1. Number of cameras.
2. Resolution of back cameras.
3. Resolution of front camera.

### Card
1. Is supported.
2. Sim Tray config
3. Size

- Some rows have OS info in the memory card column.

### OS
1. OS --> Android iOS
2. Version


**Feature Phones**

1. Low priced. Lower than Rs.5000/-.
2. Small screen size - Less than 3 inch
3. Smaller battery capicities - Lower than 2000 mAh

In [ ]:
df.sample(30)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
146,iQOO 7,"₹24,990",83.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 870, Octa Core, 3.2 GHz Processor","8 GB RAM, 128 GB inbuilt",4400 mAh Battery with 66W Fast Charging,"6.62 inches, 1080 x 2400 px, 120 Hz Display wi...",48 MP + 13 MP + 2 MP Triple Rear & 16 MP Front...,Android v11,No FM Radio
124,Samsung Galaxy M33 5G (8GB RAM + 128GB),"₹19,499",84.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Exynos 1280, Octa Core, 2.4 GHz Processor","8 GB RAM, 128 GB inbuilt",6000 mAh Battery with 25W Fast Charging,"6.6 inches, 1080 x 2400 px, 120 Hz Display wit...",50 MP Quad Rear & 8 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v12
405,Realme 9i (4GB RAM + 128GB),"₹13,999",75.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Snapdragon 680, Octa Core, 2.4 GHz Processor","4 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.6 inches, 1080 x 2412 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card Supported, upto 1 TB",Android v11
915,OPPO A15,"₹9,999",66.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio P35, Octa Core, 2.3 GHz Processor","3 GB RAM, 32 GB inbuilt",4230 mAh Battery,"6.52 inches, 720 x 1600 px Display with Water ...",13 MP + 2 MP + 2 MP Triple Rear & 5 MP Front C...,"Memory Card Supported, upto 256 GB",Android v10.0
858,Oppo Reno 5,"₹29,990",83.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 765G , Octa Core, 2.4 GHz Processor","8 GB RAM, 128 GB inbuilt",4300 mAh Battery with 65W Fast Charging,"6.43 inches, 1080 x 2400 px Display with Punch...",64 MP Quad Rear & 32 MP Front Camera,Android v11,Bluetooth
316,Vivo Y100,"₹29,990",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 900, Octa Core, 2.4 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 66W Fast Charging,"6 inches, 1080 x 2400 px, 120 Hz Display with ...",50 MP + 8 MP + 2 MP Triple Rear & 50 MP + 8 MP...,Memory Card Not Supported,Android v12
368,Samsung Galaxy S21 FE 5G (8GB RAM + 256GB),"₹43,999",88.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Exynos 2100, Octa Core, 2.9 GHz Processor","8 GB RAM, 256 GB inbuilt",4500 mAh Battery with 25W Fast Charging,"6.4 inches, 1080 x 2340 px, 120 Hz Display wit...",12 MP + 12 MP + 8 MP Triple Rear & 32 MP Front...,Memory Card Not Supported,Android v11
997,Poco C55,"₹7,999",70.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio G85, Octa Core, 2 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery,"6.71 inches, 720 x 1650 px Display with Water ...",50 MP + Depth Sensor Dual Rear & 5 MP Front Ca...,"Memory Card Supported, upto 512 GB",Android v12
879,Nokia X30,"₹41,990",85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",4200 mAh Battery with 33W Fast Charging,"6.43 inches, 1080 x 2400 px, 90 Hz Display wit...",50 MP + 13 MP Dual Rear & 16 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v12
219,Samsung Galaxy A04e,"₹9,299",65.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio P35, Octa Core, 2.3 GHz Processor","3 GB RAM, 32 GB inbuilt",5000 mAh Battery,"6.5 inches, 720 x 1600 px Display with Water D...",13 MP + 2 MP Dual Rear & 5 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v12


In [ ]:
# remove feature phones from data

def extract_feature_phones(df: pd.DataFrame):
    temp_df = (
        df
        .assign(
            price=lambda df_: (
                df_["price"]
                .str.replace("₹", "")
                .str.replace(",", "")
                .astype(np.float64)
            ),
            battery = lambda df_: (
                df_["battery"]
                .str.extract("(\d+).?mAh")
                .astype(np.float64)
        ),
            display = lambda df_:(
                df_["display"]
                .str.extract("(\d+\.\d*).?inches")
                .astype(np.float64)
            )
        )
    )

    battery_cond = temp_df["battery"] <= 2000
    display_cond = temp_df["display"] <= 4.0
    price_cond = temp_df["price"] <= 4000


    return (
        temp_df
        .loc[(battery_cond) | (display_cond) | (price_cond)]
    )


extract_feature_phones(df)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
155,Nokia 2780 Flip,4990.0,NaN,"Dual Sim, 3G, 4G, Wi-Fi","Snapdragon QM215, Quad Core, 1.3 GHz Processor","4 GB RAM, 512 MB inbuilt",1450.0,2.70,Dual Display,5 MP Rear Camera,"Memory Card Supported, upto 32 GB"
190,Realme C2s,3499.0,63.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio P22 , Octa Core, 2 GHz Processor","3 GB RAM, 32 GB inbuilt",4000.0,6.10,13 MP + 2 MP Dual Rear & 5 MP Front Camera,"Memory Card Supported, upto 128 GB",Android v9.0 (Pie)
191,Duoqin F22 Pro,9990.0,NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio G85, Octa Core, 2 GHz Processor","4 GB RAM, 64 GB inbuilt",2150.0,3.54,8 MP Rear & 2 MP Front Camera,Memory Card Supported,Android v12
271,Nokia 2720 V Flip,6199.0,NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Snapdragon 205 , Dual Core, 1.1 GHz Processor","512 MB RAM, 4 GB inbuilt",1500.0,2.80,Dual Display,2 MP Rear Camera,Memory Card Supported
400,Jio JioPhone 2,2999.0,NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Dual Core, 1 GHz Processor","512 MB RAM, 4 GB inbuilt",2000.0,2.40,2 MP Rear & 0.3 MP Front Camera,"Memory Card Supported, upto 128 GB",KAI OS
473,Nokia 110 4G,1762.0,NaN,"Dual Sim, 3G, 4G, VoLTE",No Wifi,"128 MB RAM, 48 MB inbuilt",1020.0,1.80,0.3 MP Rear Camera,"Memory Card Supported, upto 32 GB",NaN
477,Vivo Y25,7499.0,NaN,"Dual Sim, 3G, 4G, Wi-Fi","Helio MT6580, Quad Core, 1.3 GHz Processor","1 GB RAM, 16 GB inbuilt",1900.0,4.50,5 MP Rear & 2 MP Front Camera,"Memory Card Supported, upto 128 GB",Android v5.1 (Lollipop)
478,Lyf Earth 1,3990.0,60.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Snapdragon 615, Octa Core, 1.5 GHz Processor","3 GB RAM, 32 GB inbuilt",3500.0,5.50,13 MP + 2 MP Dual Rear & 5 MP Front Camera,"Memory Card (Hybrid), upto 64 GB",Android v5.1.1 (Lollipop)
486,Nokia 8000 4G,6899.0,NaN,"Dual Sim, 3G, 4G, Wi-Fi","Snapdragon 210, Quad Core, 1.1 GHz Processor","512 MB RAM, 4 GB inbuilt",1500.0,2.80,2 MP Rear Camera,Memory Card Supported,Bluetooth
532,Samsung Guru Music 2 Dual Sim,1949.0,NaN,Dual Sim,No Wifi,"Single Core, 208 MHz Processor",800.0,NaN,No Rear Camera,"Memory Card Supported, upto 16 GB",NaN


In [ ]:
df

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,OnePlus 11 5G,"₹54,999",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,Samsung Galaxy A14 5G,"₹16,499",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13
3,Motorola Moto G62 5G,"₹14,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
4,Realme 10 Pro Plus,"₹24,999",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...,Memory Card Not Supported,Android v13
...,...,...,...,...,...,...,...,...,...,...,...
1015,Motorola Moto Edge S30 Pro,"₹34,990",83.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 8 Gen1, Octa Core, 3 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 68.2W Fast Charging,"6.67 inches, 1080 x 2460 px, 120 Hz Display wi...",64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,No FM Radio
1016,Honor X8 5G,"₹14,990",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 480+, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 22.5W Fast Charging,"6.5 inches, 720 x 1600 px Display with Water D...",48 MP + 2 MP + Depth Sensor Triple Rear & 8 MP...,"Memory Card Supported, upto 1 TB",Android v11
1017,POCO X4 GT 5G (8GB RAM + 256GB),"₹28,990",85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Dimensity 8100, Octa Core, 2.85 GHz Processor","8 GB RAM, 256 GB inbuilt",5080 mAh Battery with 67W Fast Charging,"6.6 inches, 1080 x 2460 px, 144 Hz Display wit...",64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Memory Card Not Supported,Android v12
1018,Motorola Moto G91 5G,"₹19,990",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.8 inches, 1080 x 2400 px Display with Punch ...",108 MP + 8 MP + 2 MP Triple Rear & 32 MP Front...,"Memory Card Supported, upto 1 TB",Android v12


In [ ]:
feature_phones_index = extract_feature_phones(df).index.to_list()

In [ ]:
print(feature_phones_index)
print(len(feature_phones_index))

[155, 190, 191, 271, 400, 473, 477, 478, 486, 532, 533, 551, 573, 575, 582, 608, 611, 627, 640, 647, 656, 657, 665, 685, 699, 748, 757, 817, 835, 844, 846, 852, 873, 881, 907, 917, 925, 926, 927, 930, 932, 988, 1000]
43


In [ ]:
def incomplete_data_phones(data: pd.DataFrame):

    columns_list = ["processor", "ram", "display"]

    temp_df = (
        data
        .assign(
            **{
                col: data[col].str.count(",")
                for col in columns_list
            }
        )
    )

    processor_split = 2
    ram_split = 1
    display_split = 2

    return (
        data
        .loc[(temp_df["processor"].lt(processor_split)) | \
        (temp_df["ram"].lt(ram_split)) | \
        (temp_df["display"].lt(display_split))]
    )



incomplete_data_phones(df)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
6,Apple iPhone 14,"₹65,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","6 GB RAM, 128 GB inbuilt",3279 mAh Battery with Fast Charging,"6.1 inches, 1170 x 2532 px Display with Small ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v16
15,Apple iPhone 13,"₹62,999",79.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","4 GB RAM, 128 GB inbuilt",3240 mAh Battery with Fast Charging,"6.1 inches, 1170 x 2532 px Display with Small ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v15
16,Vivo Y16,"₹9,999",65.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio P35, Octa Core, 2.3 GHz Processor","3 GB RAM, 32 GB inbuilt",5000 mAh Battery with 10W Fast Charging,"6.51 inches, 720 x 1600 px Display with Water ...",13 MP + 2 MP Dual Rear & 5 MP Front Camera,Memory Card Supported,Android v12
19,Vivo Y22,"₹14,499",72.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio G70, Octa Core, 2 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 18W Fast Charging,"6.55 inches, 720 x 1612 px Display with Water ...",50 MP + 2 MP Dual Rear & 8 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v12
27,Apple iPhone 14 Pro Max,"₹1,29,990",76.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A16, Hexa Core Processor","6 GB RAM, 128 GB inbuilt",4323 mAh Battery with Fast Charging,"6.7 inches, 1290 x 2796 px, 120 Hz Display",48 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v16
...,...,...,...,...,...,...,...,...,...,...,...
1012,itel A23s,"₹4,787",NaN,"Dual Sim, 3G, 4G, Wi-Fi","Spreadtrum SC9832E, Quad Core, 1.4 GHz Processor","2 GB RAM, 32 GB inbuilt",3020 mAh Battery,"5 inches, 854 x 480 px Display",2 MP Rear Camera,Android v11,No FM Radio
1013,Google Pixel 8 Pro,"₹70,990",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Google Tensor 3, Octa Core Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.73 inches, 1440 x 3120 px, 120 Hz Display wi...",50 MP + 50 MP + 50 MP Triple Rear & 12 MP Fron...,Android v13,No FM Radio
1016,Honor X8 5G,"₹14,990",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 480+, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 22.5W Fast Charging,"6.5 inches, 720 x 1600 px Display with Water D...",48 MP + 2 MP + Depth Sensor Triple Rear & 8 MP...,"Memory Card Supported, upto 1 TB",Android v11
1018,Motorola Moto G91 5G,"₹19,990",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.8 inches, 1080 x 2400 px Display with Punch ...",108 MP + 8 MP + 2 MP Triple Rear & 32 MP Front...,"Memory Card Supported, upto 1 TB",Android v12


In [ ]:
df.head(3)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,OnePlus 11 5G,"₹54,999",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,Samsung Galaxy A14 5G,"₹16,499",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13


In [ ]:
import re

In [ ]:
def get_apple_phones(data: pd.DataFrame):
    return (
        data
        .loc[(
            data["model"]
            .str.contains("apple", flags=re.IGNORECASE)
        ), :]
    )

get_apple_phones(df)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
6,Apple iPhone 14,"₹65,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","6 GB RAM, 128 GB inbuilt",3279 mAh Battery with Fast Charging,"6.1 inches, 1170 x 2532 px Display with Small ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v16
15,Apple iPhone 13,"₹62,999",79.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","4 GB RAM, 128 GB inbuilt",3240 mAh Battery with Fast Charging,"6.1 inches, 1170 x 2532 px Display with Small ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v15
27,Apple iPhone 14 Pro Max,"₹1,29,990",76.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A16, Hexa Core Processor","6 GB RAM, 128 GB inbuilt",4323 mAh Battery with Fast Charging,"6.7 inches, 1290 x 2796 px, 120 Hz Display",48 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v16
33,Apple iPhone 11,"₹38,999",73.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","A13 Bionic, Hexa Core, 2.65 GHz Processor","4 GB RAM, 64 GB inbuilt",3110 mAh Battery,"6.1 inches, 828 x 1792 px Display with Large N...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v13
56,Apple iPhone 14 Plus,"₹74,999",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","6 GB RAM, 128 GB inbuilt",4325 mAh Battery with Fast Charging,"6.7 inches, 1284 x 2778 px Display with Small ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v16
76,Apple iPhone 13 Pro,"₹1,19,900",83.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","6 GB RAM, 128 GB inbuilt",3095 mAh Battery with Fast Charging,"6.1 inches, 1170 x 2532 px, 120 Hz Display wit...",12 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v15.0
100,Apple iPhone 14 Pro,"₹1,19,990",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A16, Hexa Core Processor","6 GB RAM, 128 GB inbuilt",3200 mAh Battery with Fast Charging,"6.1 inches, 1179 x 2556 px, 120 Hz Display",48 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v16
111,Apple iPhone 12,"₹51,999",74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
136,Apple iPhone 11 (128GB),"₹46,999",75.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","A13 Bionic, Hexa Core, 2.65 GHz Processor","4 GB RAM, 128 GB inbuilt",3110 mAh Battery,"6.1 inches, 828 x 1792 px Display with Large N...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v13
149,Apple iPhone 12 Mini,"₹40,999",74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio


In [ ]:
get_apple_phones(df).shape

(47, 11)

In [ ]:
def get_ipods(df: pd.DataFrame):
  return (
      df
      .loc[(
          df["model"]
          .str.contains("iPod", flags=re.IGNORECASE)
      ), :]
  )

get_ipods(df)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
754,Apple iPod Touch (7th Gen),"₹18,900",NaN,Wi-Fi,32 GB inbuilt,"4 inches, 640 x 1136 px Display",8 MP Rear & 1.2 MP Front Camera,iOS v12,No FM Radio,Bluetooth,Browser


In [ ]:
~df.sample(10)["display"].str.contains("inches", flags=re.IGNORECASE)

,display
892,False
58,False
454,False
219,False
933,False
611,True
33,False
27,False
758,False
930,True


In [ ]:
def get_apple_shifted_phones(data: pd.DataFrame):
    apple_data = get_apple_phones(data)

    return (
        apple_data
        .loc[
            (~apple_data["display"].str.contains("inches", flags=re.IGNORECASE)) | \
            (~apple_data["camera"].str.contains("MP", flags=re.IGNORECASE)) | \
            (~apple_data["card"].str.contains("Card", flags=re.IGNORECASE))
        ]
    )


get_apple_shifted_phones(df)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
111,Apple iPhone 12,"₹51,999",74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
149,Apple iPhone 12 Mini,"₹40,999",74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt","5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
307,Apple iPhone 12 (128GB),"₹55,999",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 128 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
363,Apple iPhone 12 Mini (128GB),"₹45,999",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 128 GB inbuilt","5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14,No FM Radio
439,Apple iPhone SE 3 2022,"₹43,900",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor",64 GB inbuilt,"4.7 inches, 750 x 1334 px Display",12 MP Rear & 7 MP Front Camera,Memory Card Not Supported,iOS v15,No FM Radio
448,Apple iPhone 15 Pro,"₹1,30,990",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC",Bionic A16,"8 GB RAM, 128 GB inbuilt","6.06 inches, 1170 x 2532 px, 120 Hz Display wi...",50 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v15,No FM Radio
607,Apple iPhone 7s,"₹52,990",NaN,"Single Sim, 3G, 4G, VoLTE, Wi-Fi","Fusion APL1024, Quad Core, 2.37 GHz Processor","3 GB RAM, 32 GB inbuilt",2230 mAh Battery,"4.7 inches, 750 x 1334 px Display",13 MP Rear & 7 MP Front Camera,iOS v10,No FM Radio
628,Apple iPhone 12 Pro (512GB),"₹1,39,900",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","6 GB RAM, 512 GB inbuilt","6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v14.0,No FM Radio
656,Apple iPhone SE 2020,"₹39,900",63.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","A13 Bionic, Hexa Core, 2.65 GHz Processor","3 GB RAM, 64 GB inbuilt",1821 mAh Battery with 18W Fast Charging,"4.7 inches, 750 x 1334 px Display",12 MP Rear & 7 MP Front Camera,iOS v13,No FM Radio
754,Apple iPod Touch (7th Gen),"₹18,900",NaN,Wi-Fi,32 GB inbuilt,"4 inches, 640 x 1136 px Display",8 MP Rear & 1.2 MP Front Camera,iOS v12,No FM Radio,Bluetooth,Browser


In [ ]:
# remove feature phones from data

def remove_feature_phones(data: pd.DataFrame):

    temp_df = (
        data
        .drop(index=extract_feature_phones(data).index.to_list())
        .drop(index=get_ipods(data).index.to_list())
    )

    return temp_df

In [ ]:
df_without_feature_phones = remove_feature_phones(df)

df_without_feature_phones.head()

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,OnePlus 11 5G,"₹54,999",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,Samsung Galaxy A14 5G,"₹16,499",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13
3,Motorola Moto G62 5G,"₹14,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
4,Realme 10 Pro Plus,"₹24,999",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...,Memory Card Not Supported,Android v13


In [ ]:
df_without_feature_phones.shape

(976, 11)

In [ ]:
# duplicates

df_without_feature_phones.duplicated().sum()

np.int64(0)

In [ ]:
# missing values

df_without_feature_phones.isna().sum()

df_without_feature_phones.loc[df_without_feature_phones["os"].isna()]

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
645,Nokia 2760 Flip,"₹5,490",NaN,"Dual Sim, 3G, 4G, Wi-Fi",1450 mAh Battery,"3.6 inches, 240 x 320 px Display",5 MP Rear & 5 MP Front Camera,"Memory Card Supported, upto 32 GB",Kaios v3.0,Bluetooth,NaN
857,LG Folder 2,"₹11,999",NaN,"Single Sim, 3G, 4G, Wi-Fi","1 GB RAM, 8 GB inbuilt",1470 mAh Battery,"2.8 inches, 240 x 320 px Display",2 MP Rear Camera,Memory Card Supported,Bluetooth,NaN


In [ ]:
def get_shifted_phones(data: pd.DataFrame):

    return (
        data
        .loc[
            (~data["display"].str.contains("inches", flags=re.IGNORECASE)) | \
            (~data["camera"].str.contains("MP", flags=re.IGNORECASE)) | \
            (~data["card"].str.contains("Card", flags=re.IGNORECASE))
        ]
    )

In [ ]:
get_shifted_phones(df_without_feature_phones)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
8,Nothing Phone 1,"₹26,749",85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 778G Plus, Octa Core, 2.5 GHz Proce...","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 33W Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 50 MP Dual Rear & 16 MP Front Camera,Android v12,No FM Radio
9,OnePlus Nord 2T 5G,"₹28,999",84.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 1300, Octa Core, 3 GHz Processor","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 80W Fast Charging,"6.43 inches, 1080 x 2400 px, 90 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Android v12,No FM Radio
12,Xiaomi Redmi Note 12 Pro 5G,"₹24,762",79.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.67 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,No FM Radio
17,OPPO Reno 9 Pro Plus,"₹45,999",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","16 GB RAM, 256 GB inbuilt",4700 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Android v13,No FM Radio
18,OnePlus 10R 5G,"₹32,999",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 8100 Max, Octa Core, 2.85 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,Bluetooth
...,...,...,...,...,...,...,...,...,...,...,...
1011,Oppo Find X6,"₹69,990",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 256 GB inbuilt",4700 mAh Battery with 120W Fast Charging,"6.73 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 48 MP + 32 MP Triple Rear & 32 MP Fron...,Android v12,No FM Radio
1012,itel A23s,"₹4,787",NaN,"Dual Sim, 3G, 4G, Wi-Fi","Spreadtrum SC9832E, Quad Core, 1.4 GHz Processor","2 GB RAM, 32 GB inbuilt",3020 mAh Battery,"5 inches, 854 x 480 px Display",2 MP Rear Camera,Android v11,No FM Radio
1013,Google Pixel 8 Pro,"₹70,990",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Google Tensor 3, Octa Core Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.73 inches, 1440 x 3120 px, 120 Hz Display wi...",50 MP + 50 MP + 50 MP Triple Rear & 12 MP Fron...,Android v13,No FM Radio
1014,Vivo X Fold 2,"₹1,19,990",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",4800 mAh Battery with 66W Fast Charging,"8.03 inches, 1916 x 2160 px, 120 Hz Display",Foldable Display,50 MP + 50 MP + 10 MP Triple Rear & 32 MP Fron...,Android v12


In [ ]:
def get_left_shifted_phones(data: pd.DataFrame):

    return (
        data
        .loc[
            ~((data["os"].str.contains("Android", flags=re.IGNORECASE)) | \
             (data["os"].str.contains("iOS", flags=re.IGNORECASE)))
        ]
    )

get_left_shifted_phones(df_without_feature_phones)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
8,Nothing Phone 1,"₹26,749",85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 778G Plus, Octa Core, 2.5 GHz Proce...","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 33W Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 50 MP Dual Rear & 16 MP Front Camera,Android v12,No FM Radio
9,OnePlus Nord 2T 5G,"₹28,999",84.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 1300, Octa Core, 3 GHz Processor","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 80W Fast Charging,"6.43 inches, 1080 x 2400 px, 90 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Android v12,No FM Radio
12,Xiaomi Redmi Note 12 Pro 5G,"₹24,762",79.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.67 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,No FM Radio
17,OPPO Reno 9 Pro Plus,"₹45,999",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","16 GB RAM, 256 GB inbuilt",4700 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Android v13,No FM Radio
18,OnePlus 10R 5G,"₹32,999",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 8100 Max, Octa Core, 2.85 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,Bluetooth
...,...,...,...,...,...,...,...,...,...,...,...
1009,Xiaomi Civi 3,"₹32,990",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Dimensity 8200, Octa Core, 3.1 GHz Processor","8 GB RAM, 256 GB inbuilt",5000 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2400 px, 120 Hz Display wit...",64 MP + 20 MP + 2 MP Triple Rear & 32 MP + 32 ...,Android v13,No FM Radio
1011,Oppo Find X6,"₹69,990",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 256 GB inbuilt",4700 mAh Battery with 120W Fast Charging,"6.73 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 48 MP + 32 MP Triple Rear & 32 MP Fron...,Android v12,No FM Radio
1012,itel A23s,"₹4,787",NaN,"Dual Sim, 3G, 4G, Wi-Fi","Spreadtrum SC9832E, Quad Core, 1.4 GHz Processor","2 GB RAM, 32 GB inbuilt",3020 mAh Battery,"5 inches, 854 x 480 px Display",2 MP Rear Camera,Android v11,No FM Radio
1013,Google Pixel 8 Pro,"₹70,990",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Google Tensor 3, Octa Core Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.73 inches, 1440 x 3120 px, 120 Hz Display wi...",50 MP + 50 MP + 50 MP Triple Rear & 12 MP Fron...,Android v13,No FM Radio


**Subset of Data**:

Things to remove:
1. feature phones & ipods --> completed
2. shifted phones
3. incomplete data

In [ ]:
indexes_of_shifted_phones = get_shifted_phones(df_without_feature_phones).index.to_list()

len(indexes_of_shifted_phones)

251

In [ ]:
indexes_of_incomplete_data = incomplete_data_phones(df_without_feature_phones).index.to_list()

len(indexes_of_incomplete_data)

390

In [ ]:
indexes_to_drop = set(indexes_of_shifted_phones)

In [ ]:
len(indexes_to_drop)

251

In [ ]:
final_df = df_without_feature_phones.drop(index=indexes_to_drop)

final_df.head()

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
0,OnePlus 11 5G,"₹54,999",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...,Memory Card Not Supported,Android v13
1,OnePlus Nord CE 2 Lite 5G,"₹19,989",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi...",64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
2,Samsung Galaxy A14 5G,"₹16,499",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Exynos 1330, Octa Core, 2.4 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.6 inches, 1080 x 2408 px, 90 Hz Display with...",50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...,"Memory Card Supported, upto 1 TB",Android v13
3,Motorola Moto G62 5G,"₹14,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 695, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,"Memory Card (Hybrid), upto 1 TB",Android v12
4,Realme 10 Pro Plus,"₹24,999",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...,Memory Card Not Supported,Android v13


In [ ]:
final_df.shape

(725, 11)

In [ ]:
725 / 1020

0.7107843137254902

## Column Wise Cleaning

### Model

In [ ]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

In [ ]:
final_df["model"]

,model
0,OnePlus 11 5G
1,OnePlus Nord CE 2 Lite 5G
2,Samsung Galaxy A14 5G
3,Motorola Moto G62 5G
4,Realme 10 Pro Plus
...,...
1010,Realme Narzo 50i Prime (4GB RAM + 64GB)
1016,Honor X8 5G
1017,POCO X4 GT 5G (8GB RAM + 256GB)
1018,Motorola Moto G91 5G


In [ ]:
# separated out brand and model_name from model

(
    final_df["model"]
    .str.split(" ", n=1, expand=True)
    .set_axis(labels=["brand", "model_name"], axis=1)
    .assign(
        model_name=lambda df_: df_["model_name"].str.replace("\(.+\)", "", regex=True)
        )
)

,brand,model_name
0,OnePlus,11 5G
1,OnePlus,Nord CE 2 Lite 5G
2,Samsung,Galaxy A14 5G
3,Motorola,Moto G62 5G
4,Realme,10 Pro Plus
...,...,...
1010,Realme,Narzo 50i Prime
1016,Honor,X8 5G
1017,POCO,X4 GT 5G
1018,Motorola,Moto G91 5G


In [ ]:
(
    final_df["model"]
    .str.split(" ", n=1, expand=True)
    .set_axis(labels=["brand", "model_name"], axis=1)
    .assign(
        model_name=lambda df_: df_["model_name"].str.replace("\(.+\)", "", regex=True)
        )
    .loc[:, "brand"]
    .str.title()
    .unique()
)

array(['Oneplus', 'Samsung', 'Motorola', 'Realme', 'Apple', 'Xiaomi',
       'Oppo', 'Vivo', 'Poco', 'Jio', 'Iqoo', 'Gionee', 'Tesla', 'Google',
       'Infinix', 'Cola', 'Tecno', 'Ikall', 'Nokia', 'Lava', 'Redmi',
       'Asus', 'Itel', 'Sony', 'Oukitel', 'Vertu', 'Blu', 'Huawei',
       'Micromax', 'Lg', 'Honor', 'Leitz', 'Doogee', 'Zte', 'Tcl',
       'Sharp', 'Blackview'], dtype=object)

In [ ]:
(
    final_df["model"]
    .str.split(" ", n=1, expand=True)
    .set_axis(labels=["brand", "model_name"], axis=1)
    .assign(
        model_name=lambda df_: df_["model_name"].str.replace("\(.+\)", "", regex=True)
        )
    .loc[:,"model_name"]
    .str.contains("[\(\)]", regex=True)
    .sum()
)

np.int64(0)

In [ ]:
def clean_model(ser:pd.Series) -> pd.DataFrame:
    return (
            ser
            .str.split(" ", n=1, expand=True)
            .set_axis(labels=["brand", "model_name"], axis=1)
            .assign(
                    model_name=lambda df_: (df_["model_name"]
                                            .str.replace("\(.+\)", "", regex=True)
                                            .str.strip()),
                    brand=lambda df_: df_["brand"].str.title().str.strip().str.title()
                    )
    )

In [ ]:
clean_model(final_df["model"])

,brand,model_name
0,Oneplus,11 5G
1,Oneplus,Nord CE 2 Lite 5G
2,Samsung,Galaxy A14 5G
3,Motorola,Moto G62 5G
4,Realme,10 Pro Plus
...,...,...
1010,Realme,Narzo 50i Prime
1016,Honor,X8 5G
1017,Poco,X4 GT 5G
1018,Motorola,Moto G91 5G


In [ ]:
clean_model(final_df["model"]).loc[:,"brand"].unique()

array(['Oneplus', 'Samsung', 'Motorola', 'Realme', 'Apple', 'Xiaomi',
       'Oppo', 'Vivo', 'Poco', 'Jio', 'Iqoo', 'Gionee', 'Tesla', 'Google',
       'Infinix', 'Cola', 'Tecno', 'Ikall', 'Nokia', 'Lava', 'Redmi',
       'Asus', 'Itel', 'Sony', 'Oukitel', 'Vertu', 'Blu', 'Huawei',
       'Micromax', 'Lg', 'Honor', 'Leitz', 'Doogee', 'Zte', 'Tcl',
       'Sharp', 'Blackview'], dtype=object)

### Price

In [ ]:
final_df["price"]

,price
0,"₹54,999"
1,"₹19,989"
2,"₹16,499"
3,"₹14,999"
4,"₹24,999"
...,...
1010,"₹8,720"
1016,"₹14,990"
1017,"₹28,990"
1018,"₹19,990"


In [ ]:
def clean_price(ser):

    return (
        ser
        .str.replace("₹", "")
        .str.replace(",", "")
        .astype(np.float64)
    )

In [ ]:
type(clean_price(final_df["price"].sample(10)))

pandas.core.series.Series

### Rating

In [ ]:
final_df["rating"]

,rating
0,89.0
1,81.0
2,75.0
3,81.0
4,82.0
...,...
1010,64.0
1016,75.0
1017,85.0
1018,80.0


In [ ]:
# missing values in rating

final_df["rating"].isna().sum()

np.int64(49)

In [ ]:
final_df[final_df["rating"].isna()]

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
14,Samsung Galaxy S23 Ultra 5G,"₹1,14,990",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 256 GB inbuilt",5000 mAh Battery with 45W Fast Charging,"6.8 inches, 1440 x 3088 px, 120 Hz Display wit...",200 MP Quad Rear & 12 MP Front Camera,Memory Card Not Supported,Android v13
49,Samsung Galaxy A74 5G,"₹42,999",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 7 Gen1, Octa Core, 2.36 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",108 MP Quad Rear & 32 MP Front Camera,"Memory Card (Hybrid), upto 1 TB",Android v12
75,Gionee G13 Pro,"₹6,190",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Tiger T310, Quad Core, 2 GHz Processor","4 GB RAM, 32 GB inbuilt",3500 mAh Battery,"6.26 inches, 720 x 1600 px Display with Large ...",13 MP Dual Rear & 5 MP Front Camera,Memory Card Supported,HarmonyOS v2
89,Samsung Galaxy A73 5G,"₹41,999",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 778G, Octa Core, 2.4 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 25W Fast Charging,"6.7 inches, 1080 x 2400 px, 120 Hz Display wit...",108 MP Quad Rear & 32 MP Front Camera,"Memory Card (Hybrid), upto 1 TB",Android v12
105,Xiaomi 13 Pro 5G,"₹58,990",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 128 GB inbuilt",4820 mAh Battery with 120W Fast Charging,"6.73 inches, 1440 x 3200 px, 120 Hz Display wi...",50.3 MP + 50 MP + 50 MP Triple Rear & 32 MP Fr...,Memory Card Not Supported,Android v13
114,Xiaomi Redmi A1,"₹6,171",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio A22, Quad Core, 2 GHz Processor","2 GB RAM, 32 GB inbuilt",5000 mAh Battery with 10W Fast Charging,"6.52 inches, 720 x 1600 px Display with Water ...",8 MP Dual Rear & 5 MP Front Camera,"Memory Card Supported, upto 512 GB",Android v12
143,Jio Phone 3,"₹4,499",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Quad Core, 1.4 GHz Processor","2 GB RAM, 64 GB inbuilt",2800 mAh Battery,"5 inches, 720 x 1280 px Display",5 MP Rear & 2 MP Front Camera,"Memory Card Supported, upto 128 GB",Android v8.1 (Oreo)
156,Realme C30,"₹5,299",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Tiger T612, Octa Core, 1.82 GHz Processor","2 GB RAM, 32 GB inbuilt",5000 mAh Battery,"6.5 inches, 720 x 1600 px Display with Water D...",8 MP Rear & 5 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v11
193,Nokia N73 5G,"₹46,999",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 8 Gen1, Octa Core, 3 GHz Processor","12 GB RAM, 256 GB inbuilt",6000 mAh Battery with 65W Fast Charging,"6.9 inches, 1400 x 3200 px, 120 Hz Display wit...",200 MP Penta Rear & 32 MP Front Camera,Memory Card Supported,Android v12.1
225,Vivo Y02,"₹8,999",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio, Octa Core Processor","3 GB RAM, 32 GB inbuilt",5000 mAh Battery with 10W Fast Charging,"6.51 inches, 720 x 1600 px Display with Water ...",8 MP Rear & 5 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v12


### Sim

In [ ]:
final_df["sim"]

,sim
0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC"
1,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi"
2,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi"
3,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi"
4,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi"
...,...
1010,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi"
1016,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi"
1017,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl..."
1018,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC"


In [ ]:
final_df.shape

(725, 11)

In [ ]:
# 4G phones in the data

(
    final_df["sim"]
    .str.contains("4G")
    .sum()
)

np.int64(724)

In [ ]:
# phone that is non 4G

(
    final_df["sim"]
    .str.contains("4G")
    .loc[lambda x: x == False]
)

,sim
735,False


In [ ]:
final_df.loc[735]

,735
model,Tecno Pop 5 Go
price,"₹6,999"
rating,NaN
sim,"Dual Sim, 3G, Wi-Fi"
processor,"Spreadtrum SC7731, Quad Core, 1.3 GHz Processor"
ram,"1 GB RAM, 16 GB inbuilt"
battery,4000 mAh Battery
display,"6.1 inches, 720 x 1520 px Display with Water D..."
camera,5 MP + 0.3 MP Dual Rear & 5 MP Front Camera
card,Memory Card Supported


In [ ]:
# number of sims

(
final_df["sim"]
.str.extract("(\w+) Sim",flags=re.IGNORECASE)
.value_counts()
)

,count
0,
Dual,718
Single,7


In [ ]:
(
    final_df["sim"]
    .str.extract("(\w+) Sim",flags=re.IGNORECASE)
    .squeeze()
    .loc[lambda ser: ser == "Single"]
    .index
)

Index([174, 429, 431, 457, 506, 870, 899], dtype='int64')

In [ ]:
final_df.loc[[174, 429, 431, 457, 506, 870, 899]]

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
174,Apple iPhone 9,"₹29,990",61.0,"Single Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","A13 Bionic, Hexa Core, 2.65 GHz Processor","3 GB RAM, 64 GB inbuilt",2050 mAh Battery with Fast Charging,"4.7 inches, 750 x 1334 px Display with Large N...",12 MP Rear & 7 MP Front Camera,Memory Card Not Supported,iOS v13.0
429,Nokia X50 5G,"₹34,999",76.0,"Single Sim, 3G, 4G, 5G, VoLTE, Wi-Fi",Snapdragon 775,"6 GB RAM, 64 GB inbuilt",6000 mAh Battery with 33W Fast Charging,"6.81 inches, 1080 x 2400 px Display with Punch...",108 MP Quad Rear & 32 MP Front Camera,Memory Card Supported,Android v11
431,Vertu Signature Touch,"₹6,50,000",62.0,"Single Sim, 3G, 4G, Wi-Fi, NFC","Snapdragon 801, Octa Core, 1.5 GHz Processor","2 GB RAM, 64 GB inbuilt",2275 mAh Battery,"4.7 inches, 1080 x 1920 px Display",13 MP Rear & 2.1 MP Front Camera,Memory Card Not Supported,Android v4.4.2 (KitKat)
457,Sony Xperia L5 5G,"₹15,990",73.0,"Single Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","Helio P35, Octa Core, 2.3 GHz Processor","4 GB RAM, 64 GB inbuilt",4000 mAh Battery with Fast Charging,"6.2 inches, 720 x 1680 px Display",13 MP + 5 MP + 2 MP Triple Rear & 8 MP Front C...,"Memory Card (Hybrid), upto 512 GB",Android v12
506,Google Pixel 2 XL,"₹15,990",69.0,"Single Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","Snapdragon 835, Octa Core, 2.35 GHz Processor","4 GB RAM, 128 GB inbuilt",3520 mAh Battery with Fast Charging,"6 inches, 1440 x 2880 px Display",12.2 MP Rear & 8 MP Front Camera,Memory Card Not Supported,Android v8.0 (Oreo)
870,Samsung Galaxy S10 5G,"₹78,990",86.0,"Single Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Exynos 9820, Octa Core, 2.7 GHz Processor","8 GB RAM, 256 GB inbuilt",4500 mAh Battery with 25W Fast Charging,"6.7 inches, 1440 x 3040 px Display",16 MP + 12 MP + 12 MP Triple Rear & 10 MP Fron...,"Memory Card (Hybrid), upto 512 GB",Android v9.0 (Pie)
899,Sony Xperia 10 III Lite 5G,"₹30,990",78.0,"Single Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 690 , Octa Core, 2 GHz Processor","6 GB RAM, 64 GB inbuilt",4500 mAh Battery with 30W Fast Charging,"6 inches, 1080 x 2520 px Display",12 MP + 8 MP + 8 MP Triple Rear & 8 MP Front C...,Memory Card Supported,Android v11


In [ ]:
def clean_sim(ser: pd.Series) -> pd.DataFrame:
    result_df = pd.DataFrame()

    return (
        result_df
        .assign(
            number_of_sim=ser.str.extract("(\w+) Sim",flags=re.IGNORECASE),
            has_5G=ser.str.contains("5G", flags=re.IGNORECASE),
            has_NFC=ser.str.contains("NFC", flags=re.IGNORECASE),
            has_IR_Blaster=ser.str.contains("IR Blaster", flags=re.IGNORECASE)
        )
        .assign(
            number_of_sim=lambda df_: np.where(df_["number_of_sim"] == "Dual", 2, 1)
        )
    )

In [ ]:
clean_sim(final_df["sim"]).shape

(725, 4)

In [ ]:
clean_sim(final_df["sim"]).isna().sum()

,0
number_of_sim,0
has_5G,0
has_NFC,0
has_IR_Blaster,0


### Processor

In [ ]:
(
    final_df["processor"]
)

,processor
0,"Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor"
1,"Snapdragon 695, Octa Core, 2.2 GHz Processor"
2,"Exynos 1330, Octa Core, 2.4 GHz Processor"
3,"Snapdragon 695, Octa Core, 2.2 GHz Processor"
4,"Dimensity 1080, Octa Core, 2.6 GHz Processor"
...,...
1010,"Tiger T612, Octa Core, 1.82 GHz Processor"
1016,"Snapdragon 480+, Octa Core, 2.2 GHz Processor"
1017,"Dimensity 8100, Octa Core, 2.85 GHz Processor"
1018,"Snapdragon 695, Octa Core, 2.2 GHz Processor"


In [ ]:
def clean_processor(ser: pd.Series):
    temp_df = pd.DataFrame()

    return (
        temp_df
        .assign(
            cpu_speed=ser.str.extract("(\d+\.?\d*).?GHz").astype(np.float64),
            number_of_cores=ser.str.extract("(\w+)\sCore"),
            processor_name=np.where(ser.str.count(",").eq(2), ser.str.split(",", n=1).str.get(0), np.nan)
        )
        .assign(
            number_of_cores= lambda df_: np.select(condlist=[df_["number_of_cores"].str.contains("Octa", case=False, na=False),
                                                            df_["number_of_cores"].str.contains("Hexa", case=False, na=False),
                                                            df_["number_of_cores"].str.contains("Quad", case=False, na=False),
                                                            df_["number_of_cores"].str.contains("Dual", case=False, na=False)],
                                                            choicelist=[8, 6, 4, 2],
                                                            default=1),
            processor_name=lambda df_: df_["processor_name"].str.split(" ", n=1).str.get(0)
        )
    )


In [ ]:
clean_processor(final_df["processor"]).join(final_df["processor"])

,cpu_speed,number_of_cores,processor_name,processor
0,3.20,8,Snapdragon,"Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor"
1,2.20,8,Snapdragon,"Snapdragon 695, Octa Core, 2.2 GHz Processor"
2,2.40,8,Exynos,"Exynos 1330, Octa Core, 2.4 GHz Processor"
3,2.20,8,Snapdragon,"Snapdragon 695, Octa Core, 2.2 GHz Processor"
4,2.60,8,Dimensity,"Dimensity 1080, Octa Core, 2.6 GHz Processor"
...,...,...,...,...
1010,1.82,8,Tiger,"Tiger T612, Octa Core, 1.82 GHz Processor"
1016,2.20,8,Snapdragon,"Snapdragon 480+, Octa Core, 2.2 GHz Processor"
1017,2.85,8,Dimensity,"Dimensity 8100, Octa Core, 2.85 GHz Processor"
1018,2.20,8,Snapdragon,"Snapdragon 695, Octa Core, 2.2 GHz Processor"


In [ ]:
clean_processor(final_df["processor"]).isna().sum()

,0
cpu_speed,36
number_of_cores,0
processor_name,47


In [ ]:
clean_processor(final_df["processor"]).dtypes

,0
cpu_speed,float64
number_of_cores,int64
processor_name,object


In [ ]:
(
    clean_processor(final_df["processor"])
    .loc[clean_processor(final_df["processor"]).isna().any(axis=1),:]
    .join(final_df["processor"])
)

,cpu_speed,number_of_cores,processor_name,processor
27,NaN,6,NaN,"Bionic A16, Hexa Core Processor"
78,NaN,8,NaN,"Exynos 1280, Octa Core Processor"
100,NaN,6,NaN,"Bionic A16, Hexa Core Processor"
118,NaN,8,NaN,Octa Core Processor
122,NaN,8,NaN,"Google Tensor, Octa Core Processor"
129,NaN,8,NaN,"Google Tensor, Octa Core Processor"
143,1.4,4,NaN,"Quad Core, 1.4 GHz Processor"
152,NaN,1,NaN,Bionic A16
188,1.6,4,NaN,"Quad Core, 1.6 GHz Processor"
201,2.0,8,NaN,"Octa Core, 2 GHz Processor"


### Ram

In [ ]:
final_df["ram"]

,ram
0,"12 GB RAM, 256 GB inbuilt"
1,"6 GB RAM, 128 GB inbuilt"
2,"4 GB RAM, 64 GB inbuilt"
3,"6 GB RAM, 128 GB inbuilt"
4,"6 GB RAM, 128 GB inbuilt"
...,...
1010,"4 GB RAM, 64 GB inbuilt"
1016,"6 GB RAM, 128 GB inbuilt"
1017,"8 GB RAM, 256 GB inbuilt"
1018,"6 GB RAM, 128 GB inbuilt"


In [ ]:
(
    final_df["ram"]
    .str.findall("(\d+).?(?:GB|TB)")
    .to_frame()
    .rename(columns={"ram": "memory"})
    .assign(
        ram= lambda df_: df_["memory"].str.get(0),
        storage=lambda df_: df_["memory"].str.get(1)
    )
)

,memory,ram,storage
0,"[12, 256]",12,256
1,"[6, 128]",6,128
2,"[4, 64]",4,64
3,"[6, 128]",6,128
4,"[6, 128]",6,128
...,...,...,...
1010,"[4, 64]",4,64
1016,"[6, 128]",6,128
1017,"[8, 256]",8,256
1018,"[6, 128]",6,128


In [ ]:
final_df["ram"]

,ram
0,"12 GB RAM, 256 GB inbuilt"
1,"6 GB RAM, 128 GB inbuilt"
2,"4 GB RAM, 64 GB inbuilt"
3,"6 GB RAM, 128 GB inbuilt"
4,"6 GB RAM, 128 GB inbuilt"
...,...
1010,"4 GB RAM, 64 GB inbuilt"
1016,"6 GB RAM, 128 GB inbuilt"
1017,"8 GB RAM, 256 GB inbuilt"
1018,"6 GB RAM, 128 GB inbuilt"


In [ ]:
def clean_ram(ser: pd.Series):
    temp_df = pd.DataFrame()
    return (
        temp_df
        .assign(
            memory= lambda df_: (ser.
                                 str.extract("(\d+).?(?:GB|TB)\sRAM", flags=re.IGNORECASE)
                                 .astype(np.float64)),
            storage=lambda df_: (ser
                                 .str.extract("(\d+).?(?:GB|TB)\sinbuilt", flags=re.IGNORECASE)
                                 .astype(np.float64))
                                 .replace(1.0, 1024.0)
        )

    )

In [ ]:
clean_ram(final_df["ram"])

,memory,storage
0,12.0,256.0
1,6.0,128.0
2,4.0,64.0
3,6.0,128.0
4,6.0,128.0
...,...,...
1010,4.0,64.0
1016,6.0,128.0
1017,8.0,256.0
1018,6.0,128.0


In [ ]:
clean_ram(final_df["ram"]).isna().sum()

,0
memory,1
storage,0


In [ ]:
clean_ram(final_df["ram"])["storage"].unique()

array([ 256.,  128.,   64.,   32.,  512., 1024.,   16.])

In [ ]:
clean_ram(final_df["ram"])["memory"].unique()

array([12.,  6.,  4.,  8.,  3.,  2., 16., 18., nan,  1.])

In [ ]:
final_df.loc[clean_ram(final_df["ram"])["storage"] == 1024]

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
290,Apple iPhone 14 Pro Max (1TB),"₹1,82,999",78.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A16, Hexa Core Processor","6 GB RAM, 1 TB inbuilt",4323 mAh Battery with Fast Charging,"6.7 inches, 1290 x 2796 px, 120 Hz Display",48 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v16
781,Apple iPhone 13 Pro Max (1TB),"₹1,79,900",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","6 GB RAM, 1 TB inbuilt",4352 mAh Battery with Fast Charging,"6.7 inches, 1284 x 2778 px, 120 Hz Display wit...",12 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v15
814,Apple iPhone 14 Pro (1TB),"₹1,72,999",77.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A16, Hexa Core Processor","6 GB RAM, 1 TB inbuilt",3200 mAh Battery with Fast Charging,"6.1 inches, 1179 x 2556 px, 120 Hz Display",48 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v16
961,Apple iPhone 13 Pro (1TB),"₹1,47,900",84.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor","6 GB RAM, 1 TB inbuilt",3095 mAh Battery with Fast Charging,"6.1 inches, 1170 x 2532 px, 120 Hz Display wit...",12 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v15.0


In [ ]:
final_df.loc[(
    clean_ram(final_df["ram"])
    .isna()
    .any(axis=1)
    )]

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
483,Huawei Mate 50 RS Porsche Design,"₹2,39,999",81.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC, IR Blaster","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor",512 GB inbuilt,4700 mAh Battery with 66W Fast Charging,"6.74 inches, 1212 x 2616 px, 120 Hz Display",50 MP + 48 MP + 13 MP Triple Rear & 13 MP Fron...,"Memory Card (Hybrid), upto 256 GB",Hongmeng OS v3.0


In [ ]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

### Battery

In [ ]:
def clean_battery(ser: pd.Series):
    temp_df = pd.DataFrame()

    return(
        temp_df
        .assign(
            capacity=ser.str.extract("(\d+).?mAh"),
            fast_charging=ser.str.contains("Fast Charging", flags=re.IGNORECASE),
            fast_charging_power=ser.str.extract("(\d+)\s?W")
        )

    )

In [ ]:
clean_battery(final_df["battery"]).join(final_df["battery"])

,capacity,fast_charging,fast_charging_power,battery
0,5000,True,100,5000 mAh Battery with 100W Fast Charging
1,5000,True,33,5000 mAh Battery with 33W Fast Charging
2,5000,True,15,5000 mAh Battery with 15W Fast Charging
3,5000,True,NaN,5000 mAh Battery with Fast Charging
4,5000,True,67,5000 mAh Battery with 67W Fast Charging
...,...,...,...,...
1010,5000,True,10,5000 mAh Battery with 10W Fast Charging
1016,5000,True,5,5000 mAh Battery with 22.5W Fast Charging
1017,5080,True,67,5080 mAh Battery with 67W Fast Charging
1018,5000,True,NaN,5000 mAh Battery with Fast Charging


In [ ]:
clean_battery(final_df["battery"]).isna().sum()

,0
capacity,0
fast_charging,0
fast_charging_power,181


In [ ]:
final_df.loc[clean_battery(final_df["battery"])["fast_charging_power"].isna(),["battery"]].join(clean_battery(final_df["battery"]))

,battery,capacity,fast_charging,fast_charging_power
3,5000 mAh Battery with Fast Charging,5000,True,NaN
6,3279 mAh Battery with Fast Charging,3279,True,NaN
15,3240 mAh Battery with Fast Charging,3240,True,NaN
25,4500 mAh Battery with Fast Charging,4500,True,NaN
27,4323 mAh Battery with Fast Charging,4323,True,NaN
...,...,...,...,...
1003,4000 mAh Battery,4000,False,NaN
1005,5000 mAh Battery,5000,False,NaN
1006,8000 mAh Battery,8000,False,NaN
1018,5000 mAh Battery with Fast Charging,5000,True,NaN


In [ ]:
(
    final_df["battery"]
)

,battery
0,5000 mAh Battery with 100W Fast Charging
1,5000 mAh Battery with 33W Fast Charging
2,5000 mAh Battery with 15W Fast Charging
3,5000 mAh Battery with Fast Charging
4,5000 mAh Battery with 67W Fast Charging
...,...
1010,5000 mAh Battery with 10W Fast Charging
1016,5000 mAh Battery with 22.5W Fast Charging
1017,5080 mAh Battery with 67W Fast Charging
1018,5000 mAh Battery with Fast Charging


### Display

In [ ]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

In [ ]:
final_df["display"]

,display
0,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit..."
1,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi..."
2,"6.6 inches, 1080 x 2408 px, 90 Hz Display with..."
3,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi..."
4,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit..."
...,...
1010,"6.5 inches, 720 x 1600 px Display with Water D..."
1016,"6.5 inches, 720 x 1600 px Display with Water D..."
1017,"6.6 inches, 1080 x 2460 px, 144 Hz Display wit..."
1018,"6.8 inches, 1080 x 2400 px Display with Punch ..."


In [ ]:
def clean_display(ser: pd.Series):
    temp_df = pd.DataFrame()
    return (
        temp_df
        .assign(
            display_size= ser.str.extract("(\d+\.?\d*)\sinches").astype(np.float64),
            resolution= ser.str.extract("(\d+.?x.?\d+).?(?:px)?"),
            refresh_rate= ser.str.extract("(\d+)\sHz")
        )
    )

In [ ]:
clean_display(final_df["display"]).join(final_df["display"])

,display_size,resolution,refresh_rate,display
0,6.70,1440 x 3216,120,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit..."
1,6.59,1080 x 2412,120,"6.59 inches, 1080 x 2412 px, 120 Hz Display wi..."
2,6.60,1080 x 2408,90,"6.6 inches, 1080 x 2408 px, 90 Hz Display with..."
3,6.55,1080 x 2400,120,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi..."
4,6.70,1080 x 2412,120,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit..."
...,...,...,...,...
1010,6.50,720 x 1600,NaN,"6.5 inches, 720 x 1600 px Display with Water D..."
1016,6.50,720 x 1600,NaN,"6.5 inches, 720 x 1600 px Display with Water D..."
1017,6.60,1080 x 2460,144,"6.6 inches, 1080 x 2460 px, 144 Hz Display wit..."
1018,6.80,1080 x 2400,NaN,"6.8 inches, 1080 x 2400 px Display with Punch ..."


In [ ]:
clean_display(final_df["display"]).isna().sum()

,0
display_size,0
resolution,0
refresh_rate,327


In [ ]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

### Camera

In [ ]:
final_df["camera"]

,camera
0,50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...
1,64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...
2,50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...
3,50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...
4,108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...
...,...
1010,8 MP Rear & 5 MP Front Camera
1016,48 MP + 2 MP + Depth Sensor Triple Rear & 8 MP...
1017,64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...
1018,108 MP + 8 MP + 2 MP Triple Rear & 32 MP Front...


In [ ]:
(
    final_df["camera"]
    .str.replace("\u2009","")
    .str.strip()
    .str.extract("(\d+)MP.* (\w+) Rear.* (\d+)MP Front")
    .rename(columns={k:v for k,v in enumerate(["rear_camera", "num_of_cameras", "front_camera"])})
)

,rear_camera,num_of_cameras,front_camera
0,50,Triple,16
1,64,Triple,16
2,50,Triple,13
3,50,Triple,16
4,108,Triple,16
...,...,...,...
1010,NaN,NaN,NaN
1016,48,Triple,8
1017,64,Triple,16
1018,108,Triple,32


In [ ]:
def clean_camera(ser: pd.Series):
    temp_df = pd.DataFrame()

    return(
            temp_df
            .assign(
                rear_camera= ser.str.extract("(\d+).?MP").astype(np.float64),
                front_camera=ser.str.extract("(\d+).?MP\sFront").astype(np.float64),
                num_of_cameras=ser.str.extract("(\w{4,6})\sRear"),
            )
    )

In [ ]:
clean_camera(final_df["camera"]).join(final_df["camera"])

,rear_camera,front_camera,num_of_cameras,camera
0,50.0,16.0,Triple,50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...
1,64.0,16.0,Triple,64 MP + 2 MP + 2 MP Triple Rear & 16 MP Front ...
2,50.0,13.0,Triple,50 MP + 2 MP + 2 MP Triple Rear & 13 MP Front ...
3,50.0,16.0,Triple,50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...
4,108.0,16.0,Triple,108 MP + 8 MP + 2 MP Triple Rear & 16 MP Front...
...,...,...,...,...
1010,8.0,5.0,NaN,8 MP Rear & 5 MP Front Camera
1016,48.0,8.0,Triple,48 MP + 2 MP + Depth Sensor Triple Rear & 8 MP...
1017,64.0,16.0,Triple,64 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...
1018,108.0,32.0,Triple,108 MP + 8 MP + 2 MP Triple Rear & 32 MP Front...


In [ ]:
clean_camera(final_df["camera"]).isna().sum()

,0
rear_camera,0
front_camera,17
num_of_cameras,53


In [ ]:
clean_camera(final_df["camera"]).loc[clean_camera(final_df["camera"]).isna().any(axis=1)].join(final_df["camera"])

,rear_camera,front_camera,num_of_cameras,camera
52,13.0,8.0,NaN,13 MP Rear & 8 MP Front Camera
70,8.0,5.0,NaN,8 MP Rear & 5 MP Front Camera
71,64.0,NaN,Triple,64 MP + 8 MP + 2 MP Triple Rear & 50 MP + 8 MP...
79,13.0,5.0,NaN,13 MP Rear & 5 MP Front Camera
125,64.0,NaN,Triple,64 MP + 8 MP + 2 MP Triple Rear & 50 MP + 8 MP...
...,...,...,...,...
979,13.0,5.0,NaN,13 MP Rear & 5 MP Front Camera
990,40.0,NaN,Quad,40 MP Quad Rear & 32 MP Dual Front Camera
996,13.0,5.0,NaN,13 MP Rear & 5 MP Front Camera
1005,20.0,13.0,NaN,20 MP Rear & 13 MP Front Camera


In [ ]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

### Card

In [ ]:
(
    final_df["card"].sample(20)
)

,card
352,"Memory Card Supported, upto 512 GB"
179,"Memory Card (Hybrid), upto 1 TB"
435,Memory Card Supported
76,Memory Card Not Supported
901,"Memory Card Supported, upto 256 GB"
387,Memory Card Not Supported
438,"Memory Card Supported, upto 512 GB"
36,"Memory Card Supported, upto 1 TB"
571,"Memory Card Supported, upto 512 GB"
214,"Memory Card Supported, upto 1 TB"


In [ ]:
def clean_card(ser: pd.Series):
    temp_df = pd.DataFrame()

    return (
        temp_df
        .assign(
            card_size=lambda df_: ser.str.extract("(\d+).?(?:TB|GB)").astype(np.float64).replace(1.0, 1024.0),
            has_card=lambda df_: np.where(df_["card_size"].notna(), True, False)
        )
    )

In [ ]:
clean_card(final_df["card"])

,card_size,has_card
0,NaN,False
1,1024.0,True
2,1024.0,True
3,1024.0,True
4,NaN,False
...,...,...
1010,1024.0,True
1016,1024.0,True
1017,NaN,False
1018,1024.0,True


### Os

In [ ]:
final_df["os"].unique()

array(['Android v13', 'Android v12', 'iOS v16', 'iOS v15', 'Android v11',
       'Android v10', 'iOS v13', 'HarmonyOS v2', 'iOS v15.0',
       'Android v10.0', 'Android v8.1 (Oreo)', 'iOS v13.0',
       'Android v11.0', 'Android v12.1', 'iOS v12.3',
       'Android v9.0 (Pie)', 'iOS v17', 'Android v4.4.2 (KitKat)',
       'Hongmeng OS v3.0', 'Android v8.0 (Oreo)',
       'Pragati OS (Powered by Android)', 'Harmony v2.0',
       'Hongmeng OS v4.0', 'HarmonyOS v2.0', 'Android v6.0 (Marshmallow)',
       'EMUI v12', 'Android v7.1 (Nougat)', 'Android', 'Android v9 (Pie)'],
      dtype=object)

## Semi Final step

In [ ]:
df.columns

Index(['model', 'price', 'rating', 'sim', 'processor', 'ram', 'battery',
       'display', 'camera', 'card', 'os'],
      dtype='object')

In [ ]:
final_df.columns[3:10].to_list() + ["model"]

['sim', 'processor', 'ram', 'battery', 'display', 'camera', 'card', 'model']

In [ ]:
def clean_phone_data(data: pd.DataFrame):
    return(
        data
        .join(data["model"].pipe(clean_model))      # cleaning model columnn --> brand, model_name
        .join(data["sim"].pipe(clean_sim))
        .join(data["processor"].pipe(clean_processor))
        .join(data["ram"].pipe(clean_ram))
        .join(data["battery"].pipe(clean_battery))
        .join(data["display"].pipe(clean_display))
        .join(data["camera"].pipe(clean_camera))
        .join(data["card"].pipe(clean_card))
        .drop(columns=["model"] + data.columns[3:10].to_list() + ["os"])
        .assign(
            price=lambda df_ : df_["price"].pipe(clean_price)
        )
    )

In [ ]:
final_df.sample(10)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
663,Infinix Hot 12 Pro,"₹10,850",74.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Tiger T616, Octa Core, 2 GHz Processor","6 GB RAM, 64 GB inbuilt",5000 mAh Battery with 18W Fast Charging,"6.6 inches, 720 x 1612 px, 90 Hz Display with ...",50 MP + Depth Sensor Dual Rear & 8 MP Front Ca...,"Memory Card Supported, upto 256 GB",Android v12
101,Xiaomi Redmi Note 11 Pro Max 5G,"₹19,999",83.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster","Snapdragon 778G, Octa Core, 2.4 GHz Processor","6 GB RAM, 128 GB inbuilt",5200 mAh Battery with 33W Fast Charging,"6.67 inches, 1080 x 2400 px, 120 Hz Display wi...",108 MP Quad Rear & 32 MP Front Camera,"Memory Card Supported, upto 512 GB",Android v11
292,Poco M4 Pro 5G (8GB RAM + 128GB),"₹15,999",82.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster","Dimensity 810 5G, Octa Core, 2.4 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.6 inches, 1080 x 2400 px, 90 Hz Display with...",50 MP + 8 MP Dual Rear & 16 MP Front Camera,"Memory Card (Hybrid), upto 1 TB",Android v11
695,Infinix Note 13 Pro,"₹16,999",80.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio G99, Octa Core, 2.2 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 120W Fast Charging,"6.91 inches, 1080 x 2460 px Display with Punch...",108 MP Quad Rear & 16 MP Front Camera,"Memory Card Supported, upto 256 GB",Android v12
245,Xiaomi Redmi Note 11T 5G (6GB RAM + 128GB),"₹16,999",81.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster","Dimensity 810 5G, Octa Core, 2.4 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 33W Fast Charging,"6.6 inches, 1080 x 2400 px, 90 Hz Display with...",50 MP + 8 MP Dual Rear & 16 MP Front Camera,"Memory Card (Hybrid), upto 1 TB",Android v11
771,Vivo Y15C,"₹8,499",66.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio P35, Octa Core, 2.3 GHz Processor","3 GB RAM, 32 GB inbuilt",5000 mAh Battery,"6.51 inches, 720 x 1600 px Display with Water ...",13 MP + 2 MP Dual Rear & 8 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v12
798,Samsung Galaxy S20 Plus,"₹83,000",88.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","Exynos 990, Octa Core, 2.73 GHz Processor","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 25W Fast Charging,"6.7 inches, 1440 x 3200 px, 120 Hz Display wit...",64 MP Quad Rear & 10 MP Front Camera,"Memory Card (Hybrid), upto 1 TB",Android v10
199,Samsung Galaxy M04,"₹8,499",66.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio P35, Octa Core, 2.3 GHz Processor","4 GB RAM, 64 GB inbuilt",5000 mAh Battery with 15W Fast Charging,"6.5 inches, 720 x 1600 px Display with Water D...",13 MP + 2 MP Dual Rear & 5 MP Front Camera,"Memory Card Supported, upto 1 TB",Android v12
325,Tecno Spark 9,"₹8,999",69.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi","Helio G37, Octa Core, 2.3 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery,"6.6 inches, 720 x 1600 px, 90 Hz Display with ...",13 MP Dual Rear & 8 MP Front Camera,"Memory Card Supported, upto 256 GB",Android v12
174,Apple iPhone 9,"₹29,990",61.0,"Single Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","A13 Bionic, Hexa Core, 2.65 GHz Processor","3 GB RAM, 64 GB inbuilt",2050 mAh Battery with Fast Charging,"4.7 inches, 750 x 1334 px Display with Large N...",12 MP Rear & 7 MP Front Camera,Memory Card Not Supported,iOS v13.0


In [ ]:
clean_phone_data(final_df)

,price,rating,brand,model_name,number_of_sim,has_5G,has_NFC,has_IR_Blaster,cpu_speed,number_of_cores,...,fast_charging,fast_charging_power,display_size,resolution,refresh_rate,rear_camera,front_camera,num_of_cameras,card_size,has_card
0,54999.0,89.0,Oneplus,11 5G,2,True,True,False,3.20,8,...,True,100,6.70,1440 x 3216,120,50.0,16.0,Triple,NaN,False
1,19989.0,81.0,Oneplus,Nord CE 2 Lite 5G,2,True,False,False,2.20,8,...,True,33,6.59,1080 x 2412,120,64.0,16.0,Triple,1024.0,True
2,16499.0,75.0,Samsung,Galaxy A14 5G,2,True,False,False,2.40,8,...,True,15,6.60,1080 x 2408,90,50.0,13.0,Triple,1024.0,True
3,14999.0,81.0,Motorola,Moto G62 5G,2,True,False,False,2.20,8,...,True,NaN,6.55,1080 x 2400,120,50.0,16.0,Triple,1024.0,True
4,24999.0,82.0,Realme,10 Pro Plus,2,True,False,False,2.60,8,...,True,67,6.70,1080 x 2412,120,108.0,16.0,Triple,NaN,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1010,8720.0,64.0,Realme,Narzo 50i Prime,2,False,False,False,1.82,8,...,True,10,6.50,720 x 1600,NaN,8.0,5.0,NaN,1024.0,True
1016,14990.0,75.0,Honor,X8 5G,2,True,False,False,2.20,8,...,True,5,6.50,720 x 1600,NaN,48.0,8.0,Triple,1024.0,True
1017,28990.0,85.0,Poco,X4 GT 5G,2,True,True,True,2.85,8,...,True,67,6.60,1080 x 2460,144,64.0,16.0,Triple,NaN,False
1018,19990.0,80.0,Motorola,Moto G91 5G,2,True,True,False,2.20,8,...,True,NaN,6.80,1080 x 2400,NaN,108.0,32.0,Triple,1024.0,True


## Left shifted phones

In [ ]:
get_left_shifted_phones(df).shape

(285, 11)

In [ ]:
(
    get_left_shifted_phones(df)
    .drop(index=[111, 149, 307, 363, 439, 448, 607, 628, 656, 754, 762, 853, 913, 914])
    .sample(50)
#     .loc[:,"card"]
#     .unique()
)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
881,Nokia 110 (2022),"₹1,599",NaN,Dual Sim,No Wifi,"4 MB RAM, 4 MB inbuilt",1000 mAh Battery,"1.77 inches, 120 x 160 px Display","Memory Card Supported, upto 32 GB",NaN,NaN
493,Nubia Red Magic 7S Pro,"₹60,990",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 135W Fast Charging,"6.8 inches, 1080 x 2400 px, 120 Hz Display",64 MP + 8 MP + 2 MP Triple Rear & 8 MP Front C...,Android v12,No FM Radio
740,OPPO Reno 6 Pro Plus 5G,"₹44,990",87.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 870, Octa Core, 3.2 GHz Processor","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 65W Fast Charging,"6.55 inches, 1080 x 2400 px, 90 Hz Display wit...",50 MP Quad Rear & 32 MP Front Camera,Android v11,No FM Radio
168,iQOO 9 SE 5G,"₹30,990",83.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 888 , Octa Core, 2.84 GHz Processor","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 66W Fast Charging,"6.62 inches, 1080 x 2400 px, 120 Hz Display wi...",48 MP + 13 MP + 2 MP Triple Rear & 16 MP Front...,Android v12,No FM Radio
649,Huawei Mate 40 Pro 4G,"₹69,990",NaN,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC, IR Blaster","Kirin 9000, Octa Core, 3.13 GHz Processor","8 GB RAM, 128 GB inbuilt",4400 mAh Battery with 66W Fast Charging,"6.76 inches, 1344 x 2772 px, 90 Hz Display wit...",50 MP + 20 MP + 12 MP Triple Rear & 13 MP Dual...,"Memory Card (Hybrid), upto 256 GB",Harmony v2.0
425,OnePlus 10 Pro 5G (12GB RAM + 512GB),"₹74,999",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen1, Octa Core, 3 GHz Processor","12 GB RAM, 512 GB inbuilt",5000 mAh Battery with 80W Fast Charging,"6.7 inches, 1440 x 3216 px, 120 Hz Display wit...",50 MP + 48 MP + 8 MP Triple Rear & 32 MP Front...,Android v12.1,No FM Radio
267,OnePlus Ace 2,"₹34,999",84.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 100W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v13,No FM Radio
229,OPPO Reno 8 Pro 5G,"₹45,999",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 8100-Max, Octa Core, 2.85 GHz Processor","12 GB RAM, 256 GB inbuilt",4500 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Android v12,No FM Radio
858,Oppo Reno 5,"₹29,990",83.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi","Snapdragon 765G , Octa Core, 2.4 GHz Processor","8 GB RAM, 128 GB inbuilt",4300 mAh Battery with 65W Fast Charging,"6.43 inches, 1080 x 2400 px Display with Punch...",64 MP Quad Rear & 32 MP Front Camera,Android v11,Bluetooth
380,Samsung Galaxy S21,"₹48,900",88.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Exynos 2100, Octa Core, 2.9 GHz Processor","8 GB RAM, 128 GB inbuilt",4000 mAh Battery with 25W Fast Charging,"6.2 inches, 1080 x 2400 px, 120 Hz Display wit...",64 MP + 12 MP + 12 MP Triple Rear & 10 MP Fron...,Android v10,No FM Radio


In [ ]:
(
    get_left_shifted_phones(df)
    .drop(index=[111, 149, 307, 363, 439, 448, 607, 628, 656, 754, 762, 853, 913, 914])
    # .loc[:,"card"]
    # .unique()
)

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
8,Nothing Phone 1,"₹26,749",85.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 778G Plus, Octa Core, 2.5 GHz Proce...","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 33W Fast Charging,"6.55 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 50 MP Dual Rear & 16 MP Front Camera,Android v12,No FM Radio
9,OnePlus Nord 2T 5G,"₹28,999",84.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 1300, Octa Core, 3 GHz Processor","8 GB RAM, 128 GB inbuilt",4500 mAh Battery with 80W Fast Charging,"6.43 inches, 1080 x 2400 px, 90 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Android v12,No FM Radio
12,Xiaomi Redmi Note 12 Pro 5G,"₹24,762",79.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster","Dimensity 1080, Octa Core, 2.6 GHz Processor","6 GB RAM, 128 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.67 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,No FM Radio
17,OPPO Reno 9 Pro Plus,"₹45,999",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8+ Gen1, Octa Core, 3.2 GHz Processor","16 GB RAM, 256 GB inbuilt",4700 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 32 MP Front ...,Android v13,No FM Radio
18,OnePlus 10R 5G,"₹32,999",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Dimensity 8100 Max, Octa Core, 2.85 GHz Processor","8 GB RAM, 128 GB inbuilt",5000 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2412 px, 120 Hz Display wit...",50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...,Android v12,Bluetooth
...,...,...,...,...,...,...,...,...,...,...,...
1009,Xiaomi Civi 3,"₹32,990",86.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...","Dimensity 8200, Octa Core, 3.1 GHz Processor","8 GB RAM, 256 GB inbuilt",5000 mAh Battery with 80W Fast Charging,"6.7 inches, 1080 x 2400 px, 120 Hz Display wit...",64 MP + 20 MP + 2 MP Triple Rear & 32 MP + 32 ...,Android v13,No FM Radio
1011,Oppo Find X6,"₹69,990",89.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor","8 GB RAM, 256 GB inbuilt",4700 mAh Battery with 120W Fast Charging,"6.73 inches, 1080 x 2400 px, 120 Hz Display wi...",50 MP + 48 MP + 32 MP Triple Rear & 32 MP Fron...,Android v12,No FM Radio
1012,itel A23s,"₹4,787",NaN,"Dual Sim, 3G, 4G, Wi-Fi","Spreadtrum SC9832E, Quad Core, 1.4 GHz Processor","2 GB RAM, 32 GB inbuilt",3020 mAh Battery,"5 inches, 854 x 480 px Display",2 MP Rear Camera,Android v11,No FM Radio
1013,Google Pixel 8 Pro,"₹70,990",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Google Tensor 3, Octa Core Processor","12 GB RAM, 256 GB inbuilt",5000 mAh Battery with 67W Fast Charging,"6.73 inches, 1440 x 3120 px, 120 Hz Display wi...",50 MP + 50 MP + 50 MP Triple Rear & 12 MP Fron...,Android v13,No FM Radio


In [ ]:
index_of_left_shifted_phones = (
                                get_left_shifted_phones(df_without_feature_phones)
                                .drop(index=[111, 149, 307, 363, 439, 448, 607, 628, 762, 853, 913, 914])
                                .loc[:,"card"]
                                .str.contains(pat="Android|iOS",case=False)
                                .index.to_list()
                            )

### **Apple data** --> left shifted phones

In [ ]:
apple_no_shifting = get_left_shifted_phones(get_apple_phones(df)).loc[:, :"ram"]

In [ ]:
apple_shifting = get_left_shifted_phones(get_apple_phones(df)).loc[:, "battery":].shift(periods=1, axis=1)

In [ ]:
apple_correct_data = apple_no_shifting.join(apple_shifting)

apple_correct_data

,model,price,rating,sim,processor,ram,battery,display,camera,card,os
111,Apple iPhone 12,"₹51,999",74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt",None,"6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14
149,Apple iPhone 12 Mini,"₹40,999",74.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 64 GB inbuilt",None,"5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14
307,Apple iPhone 12 (128GB),"₹55,999",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 128 GB inbuilt",None,"6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14
363,Apple iPhone 12 Mini (128GB),"₹45,999",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","4 GB RAM, 128 GB inbuilt",None,"5.4 inches, 1080 x 2340 px Display",12 MP + 12 MP Dual Rear & 12 MP Front Camera,Memory Card Not Supported,iOS v14
439,Apple iPhone SE 3 2022,"₹43,900",NaN,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A15, Hexa Core, 3.22 GHz Processor",64 GB inbuilt,None,"4.7 inches, 750 x 1334 px Display",12 MP Rear & 7 MP Front Camera,Memory Card Not Supported,iOS v15
448,Apple iPhone 15 Pro,"₹1,30,990",75.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC",Bionic A16,"8 GB RAM, 128 GB inbuilt",None,"6.06 inches, 1170 x 2532 px, 120 Hz Display wi...",50 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v15
607,Apple iPhone 7s,"₹52,990",NaN,"Single Sim, 3G, 4G, VoLTE, Wi-Fi","Fusion APL1024, Quad Core, 2.37 GHz Processor","3 GB RAM, 32 GB inbuilt",None,2230 mAh Battery,"4.7 inches, 750 x 1334 px Display",13 MP Rear & 7 MP Front Camera,iOS v10
628,Apple iPhone 12 Pro (512GB),"₹1,39,900",80.0,"Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC","Bionic A14, Hexa Core, 3.1 GHz Processor","6 GB RAM, 512 GB inbuilt",None,"6.1 inches, 1170 x 2532 px Display with Large ...",12 MP + 12 MP + 12 MP Triple Rear & 12 MP Fron...,Memory Card Not Supported,iOS v14.0
656,Apple iPhone SE 2020,"₹39,900",63.0,"Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC","A13 Bionic, Hexa Core, 2.65 GHz Processor","3 GB RAM, 64 GB inbuilt",None,1821 mAh Battery with 18W Fast Charging,"4.7 inches, 750 x 1334 px Display",12 MP Rear & 7 MP Front Camera,iOS v13
754,Apple iPod Touch (7th Gen),"₹18,900",NaN,Wi-Fi,32 GB inbuilt,"4 inches, 640 x 1136 px Display",None,8 MP Rear & 1.2 MP Front Camera,iOS v12,No FM Radio,Bluetooth


In [ ]:
apple_correct_data.isna().sum()

,0
model,0
price,0
rating,3
sim,0
processor,0
ram,0
battery,14
display,0
camera,0
card,0


# Final Cleaning Step

In [ ]:
725 + 14

739

In [ ]:
df_shifted_phones = df_without_feature_phones.loc[index_of_left_shifted_phones,:]

In [ ]:
final_data = pd.concat([final_df, apple_correct_data, df_shifted_phones])

final_data.duplicated().sum()

# drop duplicates

final_data.drop_duplicates(inplace=True)

final_data.shape

(966, 11)

In [ ]:
final_data.duplicated().sum()

np.int64(0)

In [ ]:
len(feature_phones_index)

43

In [ ]:
1020 - 43

977

In [ ]:
def clean_phone_data(data: pd.DataFrame):
    return(
        data
        .join(data["model"].pipe(clean_model))
        .join(data["sim"].pipe(clean_sim))
        .join(data["processor"].pipe(clean_processor))
        .join(data["ram"].pipe(clean_ram))
        .join(data["battery"].pipe(clean_battery))
        .join(data["display"].pipe(clean_display))
        .join(data["camera"].pipe(clean_camera))
        .join(data["card"].pipe(clean_card))
        .drop(columns=["model"] + data.columns[3:10].to_list() + ["os"])
        .assign(
            price=lambda df_ : df_["price"].pipe(clean_price)
        )
    )

In [ ]:
cleaned_data = clean_phone_data(final_data)

cleaned_data.head()

,price,rating,brand,model_name,number_of_sim,has_5G,has_NFC,has_IR_Blaster,cpu_speed,number_of_cores,...,fast_charging,fast_charging_power,display_size,resolution,refresh_rate,rear_camera,front_camera,num_of_cameras,card_size,has_card
0,54999.0,89.0,Oneplus,11 5G,2,True,True,False,3.2,8,...,True,100,6.70,1440 x 3216,120,50.0,16.0,Triple,NaN,False
1,19989.0,81.0,Oneplus,Nord CE 2 Lite 5G,2,True,False,False,2.2,8,...,True,33,6.59,1080 x 2412,120,64.0,16.0,Triple,1024.0,True
2,16499.0,75.0,Samsung,Galaxy A14 5G,2,True,False,False,2.4,8,...,True,15,6.60,1080 x 2408,90,50.0,13.0,Triple,1024.0,True
3,14999.0,81.0,Motorola,Moto G62 5G,2,True,False,False,2.2,8,...,True,NaN,6.55,1080 x 2400,120,50.0,16.0,Triple,1024.0,True
4,24999.0,82.0,Realme,10 Pro Plus,2,True,False,False,2.6,8,...,True,67,6.70,1080 x 2412,120,108.0,16.0,Triple,NaN,False


In [ ]:
# save as csv

cleaned_data.to_csv("cleaned_smartphones_data.csv", index=False)

In [ ]:
# check for missing values in clean data

cleaned_data.isna().sum()

,0
price,0
rating,91
brand,0
model_name,0
number_of_sim,0
has_5G,0
has_NFC,0
has_IR_Blaster,0
cpu_speed,47
number_of_cores,0


In [ ]:
# shape of cleaned data

cleaned_data.shape

(966, 24)